ISSUER_PROTOCOL.md

```diff
@@ -0,0 +1,343 @@
+
+struct {
+  opaque t<1..2^16-1>; // X9.62 Uncompressed point.
+  opaque s[Nn];
+};
+
+Hc(x) - P384_XMD:SHA-512_SSWU_RO_ scalar derivation (using the hash-to-curve hash_to_field operation modulo the group order rather than the field element) with the input x and a dst of "PMBTokens Experiment V1 HashC". Depending on the context, one of the following serializations is used for the input:
+
+opaque ECPoint<1..2^16-1>; // X9.62 Uncompressed point.
+
+struct {
+  uint8 label[6] = "DLEQ2\0";
+  ECPoint X;
+  ECPoint T;
+  ECPoint S;
+  ECPoint W;
+  ECPoint K0;
+  ECPoint K1;
+
+} DLEQInput; // Used for DLEQ proof.
+
+struct {
+  uint8 label[8] = "DLEQOR2\0";
+  ECPoint X0;
+  ECPoint X1;
+  ECPoint T;
+  ECPoint S;
+  ECPoint W;
+  ECPoint K00;
+  ECPoint K01;
+  ECPoint K10;
+  ECPoint K11;
+} DLEQORInput; // Used for DLEQOR proof.
+
+struct {
+  uint8 label[11] = "DLEQ Batch\0";
+  ECPoint pubs;
+  ECPoint pub0;
+  ECPoint pub1;
+  ECPoint T'0;
+  ECPoint S'0;
+  ECPoint Wp0;
+  ECPoint Wsp0;
+  ...
+  ECPoint T'n;
+  ECPoint S'n;
+  ECPoint Wpn;
+  ECPoint Wspn;
+  uint16 index;
+} DLEQBatchInput; // Used for Batch proof.
+```
+
+### Issue Function
+
+The _Issue_ function corresponds to the **AT.Sig** stage of the PMBTokens protocol.
+
+Inputs:
+
+*   count (the number of tokens being requested, from the IssueRequest)
+*   nonces (the list of nonces to sign)
+*   secretKey (the secret key that should be used to sign this request, determined by the public metadata)
+*   publicKey (the public key corresponding to the secretKey)
+*   keyID (the ID corresponding to this key)
+*   privateMetadata (the boolean value corresponding to the private metadata value)
+*   toIssue (number of tokens to issue)
+
+Outputs:
+
+*   issued (the number of tokens being issued)
+*   keyID (the key used to sign these tokens)
+*   signed (the list of signed nonces)
+*   cs, us, vs, c0, u0, v0, c1, u1, v1 (the list of scalars corresponding to the zero-knowledge proof)
+
+```
+Issue:
+  issued = min(count, toIssue)
+  xb,yb = secretKey.x1, secretKey.y1
+  if privateMetadata == 0:
+    xb,yb = secretKey.x0, secretKey.y0
+  signed = []
+  T,S,W,Ws = [], [], [], []
+  for i in 0..issued:
+    T' = nonces[i]
+    s ←$ {0, 1}λ
+    S' = Hs(T', s)
+    Wp = xb*T' + yb*S'
+    Wsp = secretKey.xs*T' + secretKey.ys*S'
+    T,S,W,Ws += T', S', Wp, Wsp
+    signed += SignedNonce{s, Wp, Wsp}
+  proof = DLEQbatched.P((publicKey,T,S,W,Ws),(secretKey.xs, secretKey.ys, xb, yb))
+  return (issued, keyID, signed, proof)
+
+DLEQbatched.P((X,T,S,W,Ws),(xs, ys, xb, yb)):
+  e0,...,em = Hc(X,T,S,W,Ws)
+  Tt = Sum(ej*Tj) // over j from 0 to m
+  St = Sum(ej*Sj) // over j from 0 to m
+  Wt = Sum(ej*Wj) // over j from 0 to m
+  Wst = Sum(ej*Wsj) // over j from 0 to m
+  dleqProof = DLEQ2.P((X,T,S,Ws), (xs,ys)) // cs,us,vs
+  dleqorProof = DLEQOR2.P((X,T,S,W), (xb,yb)) // c0,c1,u0,u1,v0,v1
+  return dleqProof + dleqorProof
+```
+
+Input Serialization:
+
+```
+Trust Token Issuance Request
```

Comment:
This ends up rendering kinda funny. Maybe outside of the code block and something like: "The whatever-whatever contains an `IssueRequest` structure defined below." That way the code is code and it's clear which of the two structures is the one you want.

(Ditto for the other section headings in code.)

Reply:
Done.